File: GMM.m
clc
close all
clear all
 
tic
 
% Import
photo = 'C:\Zephyr\MATLAB\*.jpg';
photopath  = 'C:\Zephyr\MATLAB\';
file = dir(photo);
 
for i = 1:length(file)
    Image{i} = imread([photopath, file(i).name]);
    Image_R(:,:,i) = Image{i}(:,:,1);
end
 
[x,y,z] = size(Image_R);
Image_R = reshape(Image_R,[x*y,z]);
Image_R = im2double(Image_R);
[~,N] = size(Image);
 
[x,y,z] = size(Image{1});
for i=1:N
    Image{i} = im2double(Image{i});
    Image{i} = reshape(Image{i},[x*y,z]);
end
 
 
% Notation
% m,n=size of image m??n=D
% N=num of image
% I0=Mu1(j,:) Finally
% Epsilon=threshold
% Alpha=Pi or weight
% Beta=Switch for N and N+U
% Mu1(j,:) Mu2(j,:)
% Sigma1 Sigma2
% P1=posterior probability of right images
% P0=1-P1 posterior probability of wrong images
% Time=Total iteration
% L_old=previous loglikely
% K=2 Clusters
 
% Initialization
Data = Image_R';
K=2; % Clusters
[N, D] = size(Data);
[Index,Center] = kmeans(Data,K);
 
for i = 1:K
    Mu(:,:,i) = zeros(x*y,z);
    Location = find(Index==i);
    Prior(i) = length(Location);
    for j=1:Prior(i)
        Mu(:,:,i) = Mu(:,:,i) + Image{Location(j)};
    end
end
 
P1 = zeros(N,1);
P1(Location) = 1;
P0 = ones(N,1)-P1;
Mu1 = Mu(:,:,1)/Prior(1);
Mu2 = Mu(:,:,2)/Prior(2);
COV1 = cell(1,D);
COV2 = cell(1,D);
 
for i = 1:N
    for j = 1:D
        COV1{j} = (Image{i}(j,:)-Mu1(j,:))'*(Image{i}(j,:)-Mu1(j,:));
        COV2{j} = (Image{i}(j,:)-Mu2(j,:))'*(Image{i}(j,:)-Mu2(j,:));
    end
    TCOV1 = cat(3,COV1{:});
    TCOV2 = cat(3,COV2{:});
    MCOV1 = mean(TCOV1,3);
    MCOV2 = mean(TCOV2,3);
    Num1(:,:,i) = P1(i)*MCOV1;
    Num2(:,:,i) = P0(i)*MCOV2;
end
Sigma1 = sum(Num1,3)/sum(P1)+eye(3);
Sigma2 = sum(Num2,3)/sum(P0)+eye(3);
 
Prior = Prior./sum(Prior);
Alpha = Prior(1);
Beta = 0;
 
 
% Parameters Setting
Epsilon = 1e-4;
L_old = -realmax;
 
k = 1;
for i = 1:N
    for j = 1:N-i
        dist(k) = norm(Image{i}-Image{j});
        k = k+1;
    end
end
 
 
% Uniform
distance = max(dist);
U = 1/(distance+1);
 
 
% Starting GMM
 
for w = 1:7
    
    % Estimation
    
    for i = 1:N
        for j = 1:D
            %  recalculate posterior probability
            Good(j) = Alpha*Gauss(Image{i}(j,:),Mu1(j,:),Sigma1);
            Bad(j) = (1-Alpha)*(Beta*Gauss(Image{i}(j,:),Mu2(j,:),Sigma2)+(1-Beta)*U);
        end
        P1(i) = sum(Good)/sum(Good+Bad);
    end
    P0 = ones(N,1)-P1;
    
    % Maximization
    
    for i = 1:N
        Num3(:,:,i) = P1(i)*Image{i};
        Num4(:,:,i) = P0(i)*Image{i};
    end
    Mu1 = sum(Num3,3)/sum(P1);
    Mu2 = sum(Num4,3)/sum(P0);
    
    
    for i = 1:N
        for j = 1:D
            COV1{j} = (Image{i}(j,:)-Mu1(j,:))'*(Image{i}(j,:)-Mu1(j,:));
            COV2{j} = (Image{i}(j,:)-Mu1(j,:))'*(Image{i}(j,:)-Mu1(j,:));
        end
        TCOV1 = cat(3,COV1{:});
        TCOV2 = cat(3,COV2{:});
        MCOV1 = mean(TCOV1,3);
        MCOV2 = mean(TCOV2,3);
        Num1(:,:,i) = P1(i)*MCOV1;
        Num2(:,:,i) = P0(i)*MCOV2;
    end
    Sigma1 = sum(Num1,3)/sum(P1)+eye(3);
    Sigma2 = sum(Num2,3)/sum(P0)+eye(3);
    Alpha = sum(P1)/N;
    
    
    % Check Likelihood
    
    % Test ??=0
    
    Beta = 0;
    
    for i=1:N
        for j=1:D
            %  recalculate posterior probability
            Good(j) = Alpha*Gauss(Image{i}(j,:),Mu1(j,:),Sigma1);
            Bad(j) = (1-Alpha)*(Beta*Gauss(Image{i}(j,:),Mu2(j,:),Sigma2)+(1-Beta)*U);
        end
    end
    
    Good = Good.^(1/D);
    Good = prod(Good);
    Bad = Bad.^(1/D);
    Bad = prod(Bad);
    
    
    for i = 1:N
        L0 = P1(i)*log(Good)+P0(i)*log(prod(Bad));
    end
    
    L0 = sum(L0);
    
    
    % Test ??=1
    
    Beta = 1;
    
    for i = 1:N
        for j = 1:D
            %  recalculate posterior probability
            Good(j) = Alpha*Gauss(Image{i}(j,:),Mu1(j,:),Sigma1);
            Bad(j) = (1-Alpha)*(Beta*Gauss(Image{i}(j,:),Mu2(j,:),Sigma2)+(1-Beta)*U);
        end
    end
    
    Good = Good.^(1/D);
    Good = prod(Good);
    Bad = Bad.^(1/D);
    Bad = prod(Bad);
    
    for i = 1:N
        L1 = P1(i)*log(prod(Good))+P0(i)*log(prod(Bad));
    end
    
    L1 = sum(L1);
    
    if L0 < L1
        L = L0;
        Beta = 0;
    end
    
    if L0 > L1
        L = L1;
        Beta = 1;
    end
    
    if abs((L/L_old)-1) < Epsilon
        break;
    else
        L_old = L;
    end
end
 
 
% Output Synthetic Image
 
I0 = reshape(Mu1,x,y,z);
imagesc(I0);
 
toc

File: GMM.py
from numpy import sum, prod, mean, size, shape, reshape, where, zeros, ones, matrix, eye, mat
from math import exp, sqrt, pi, log
from scipy.linalg import norm, det, pinv
from skimage import io
from sklearn.cluster import KMeans
from time import clock
 
start = clock()

def Gauss(Data, Mu, Sigma):
       
#Author:	Louis, 2017
#
#Inputs -----------------------------------------------------------------
#  o Data:  D x N matrix representing N datapoints of D dimensions.
#  o Mu:    D x K matrix representing the centers of the K GMM components.
#  o Sigma: D x D x K matrix representing the Covariance matrices of the 
#           K GMM components.
#Outputs ----------------------------------------------------------------
#  o prob:  1 x N matrix representing the probabilities for the 
#           N datapoints.    

       D = shape(Data)
       D = D[0]
       Data = Data-Mu
       power = -0.5*(Data.T@pinv(Sigma)@Data)
       prob = exp(power)/sqrt((2*pi)**D*(abs(det(Sigma))+1))
       return prob

       
path='C:\Zephyr\Python\*.jpg'
Image = io.ImageCollection(path)
(N,x,y,z)=shape(Image)
Image = reshape(Image,(N,x*y,z))
D = x*y


# Notation
#m,n=size of image m×n=D
#N=num of image
#I0=Mu1(j,:) Finally
#Epsilon=threshold
#Alpha=Pi or weight
#Beta=Switch for N and N+U
#Mu1(j,:) Mu2(j,:)
#Sigma1 Sigma2
#P1=posterior probability of right images
#P2=1-P1 posterior probability of wrong images
#Time=Total iteration
#L_old=previous loglikely
#K=2 Clusters


# Initialization
Image_R = Image[:,:,0]
Image_R = reshape(Image_R,(N,D))
K = KMeans(n_clusters=2).fit(Image_R)
Index = K.labels_
Center = K.cluster_centers_
Mu = zeros(shape=[2,D,3])
Prior = zeros(2)

for i in range(2):
       Location = matrix(where(Index==i)).T   
       Prior[i] = size(Location)
       Mu[i,:,:]= sum(Image[Location,:,:],axis=0)

P1 = Index
P2 = ones(N)-P1
Mu1 = Mu[0,:,:]/Prior[0]
Mu2 = Mu[1,:,:]/Prior[1]
Cov1 = zeros(shape=[D,3,3])
Cov2 = zeros(shape=[D,3,3])
Num1 = zeros(shape=[N,3,3])
Num2 = zeros(shape=[N,3,3])
Num3 = zeros(shape=[N,D,3])
Num4 = zeros(shape=[N,D,3])

for i in range(N):
    for j in range(D):
        Cov1[j,:,:] = mat(Image[i,j,:]-Mu1[j,:]).T*mat(Image[i,j,:]-Mu1[j,:])
        Cov2[j,:,:] = mat(Image[i,j,:]-Mu2[j,:]).T*mat(Image[i,j,:]-Mu2[j,:])      
    MCov1 = mean(Cov1,axis=0)
    MCov2 = mean(Cov2,axis=0)
    Num1[i,:,:] = P1[i]*MCov1
    Num2[i,:,:] = P2[i]*MCov2
Sigma1 = sum(Num1,axis=0)/sum(P1)+eye(3)
Sigma2 = sum(Num2,axis=0)/sum(P2)+eye(3)

Prior = len(Location)/N
Alpha = 1-Prior
Beta = 0


# Parameters Setting
Epsilon = 1e-4
L_old = 0
Good = zeros(D)
Bad = zeros(D)

# Uniform
dist = zeros(N*(N+1)//2) # note the // here to make it int
k = 0
for i in range(N):
    for j in range(N-i):
        dist[0] = norm(Image[i,:,:]-Image[j,:,:])
        k = k+1
        
distance = max(dist)
U = 1/(distance+1)


# Starting GMM

for w in range(1):  

    # Estimation
    for i in range(N):   
        for j in range(D):
            #  recalculate posterior probability
            Good[j] = Alpha*Gauss(Image[i,j,:],Mu1[j,:],Sigma1)
            Bad[j] = (1-Alpha)*(Beta*Gauss(Image[i,j,:],Mu2[j,:],Sigma2)+(1-Beta)*U)
        P1[i] = sum(Good)/sum(Good+Bad)
    P2 = ones(N)-P1
    
    
    # Maximization
    
    for i in range(N):
        Num3[i,:,:] = P1[i]*Image[i]
        Num4[i,:,:] = P2[i]*Image[i]
    Mu1 = sum(Num3,axis=0)/sum(P1)
    Mu2 = sum(Num4,axis=0)/sum(P2)
    
    for i in range(N):
           for j in range(D):
                  Cov1[j,:,:] = mat(Image[i,j,:]-Mu1[j,:]).T*mat(Image[i,j,:]-Mu1[j,:])
                  Cov2[j,:,:] = mat(Image[i,j,:]-Mu2[j,:]).T*mat(Image[i,j,:]-Mu2[j,:])        
           MCov1 = mean(Cov1,axis=0)
           MCov2 = mean(Cov2,axis=0)
           Num1[i,:,:] = P1[i]*MCov1
           Num2[i,:,:] = P2[i]*MCov2
    Sigma1 = sum(Num1,axis=0)/sum(P1)
    Sigma2 = sum(Num1,axis=0)/sum(P2)   

    Alpha = sum(P1)/N
    
    
    # Check Likelihood
    
    # Test β=0
    
    Beta = 0   
    
    for i in range(N):   
        for j in range(D):
            #  recalculate posterior probability
            Good[j] = Alpha*Gauss(Image[i,j,:],Mu1[j,:],Sigma1)
            Bad[j] = (1-Alpha)*(Beta*Gauss(Image[i,j,:],Mu2[j,:],Sigma2)+(1-Beta)*U)
    Good = Good**(1/D)
    Good = prod(Good)
    Bad = Bad**(1/D)
    Bad = prod(Bad)
    
    for i in range(N):
        L0 = P1[i]*log(Good)+P2[i]*log(prod(Bad))
      
    L0 = sum(L0)


    # Test β=1
    
    Beta = 1
    
    for i in range(N):   
        for j in range(D):
            #  recalculate posterior probability
            Good[j] = Alpha*Gauss(Image[i,j,:],Mu1[j,:],Sigma1)
            Bad[j] = (1-Alpha)*(Beta*Gauss(Image[i,j,:],Mu2[j,:],Sigma2)+(1-Beta)*U)
    
    G = Good**(1/D)
    G = prod(Good)
    B = Bad**(1/D)
    B = prod(Bad)
    
    for i in range(N):
        L1 = P1[i]*log(prod(G))+P2[i]*log(prod(B))   
    
    L1 = sum(L1)
    
    if L0 < L1:
        L = L0
        Beta = 0  
    
    if L0 > L1:
        L = L1
        Beta = 1  
    
    if abs((L/L_old)-1) < Epsilon:
        break
    else:
        L_old = L
    

# Output Synthetic Image

I0 = reshape(Mu1,x,y,3)
io.show(I0)

end = clock()
print("Time used:", end-start)

